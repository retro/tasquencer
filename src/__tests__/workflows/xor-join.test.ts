import { Effect } from 'effect';
import { expect, it } from 'vitest';

import { Builder, IdGenerator, Service } from '../../index.js';
import { getEnabledTaskNames, makeIdGenerator } from '../shared.js';

it('runs a net with "xor" join', () => {
  const workflowDefinition = Builder.workflow()
    .withName('xor-join')
    .startCondition('start')
    .task('initial_task', Builder.emptyTask())
    .condition('choice')
    .task('task_a', Builder.emptyTask())
    .task('task_b', Builder.emptyTask())
    .task('task_c', Builder.emptyTask())
    .task('finish_task', Builder.emptyTask().withJoinType('xor'))
    .endCondition('end')
    .connectCondition('start', (to) => to.task('initial_task'))
    .connectTask('initial_task', (to) => to.condition('choice'))
    .connectCondition('choice', (to) =>
      to.task('task_a').task('task_b').task('task_c')
    )
    .connectTask('task_a', (to) => to.task('finish_task'))
    .connectTask('task_b', (to) => to.task('finish_task'))
    .connectTask('task_c', (to) => to.task('finish_task'))
    .connectTask('finish_task', (to) => to.condition('end'));
  const program = Effect.gen(function* ($) {
    const idGenerator = makeIdGenerator();

    const service = yield* $(
      workflowDefinition.build(),
      Effect.flatMap((workflow) => Service.initialize(workflow)),
      Effect.provideService(IdGenerator, idGenerator)
    );

    expect(yield* $(service.getState())).toMatchSnapshot();

    yield* $(service.start());
    const state1 = yield* $(service.getState());
    expect(state1).toMatchSnapshot();
    expect(getEnabledTaskNames(state1)).toEqual(new Set(['initial_task']));

    yield* $(service.startTask('initial_task'));
    const state2 = yield* $(service.getState());
    expect(state2).toMatchSnapshot();
    expect(getEnabledTaskNames(state2)).toEqual(
      new Set(['task_a', 'task_b', 'task_c'])
    );

    yield* $(service.startTask('task_b'));
    const state3 = yield* $(service.getState());
    expect(state3).toMatchSnapshot();
    expect(getEnabledTaskNames(state3)).toEqual(new Set(['finish_task']));
  });

  Effect.runSync(program);
});
