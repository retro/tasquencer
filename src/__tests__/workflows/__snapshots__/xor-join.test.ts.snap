// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runs a net with "xor" join 1`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "choice",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "initial_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_b",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_c",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "xor-join",
      "parent": null,
      "state": "initialized",
    },
  ],
}
`;

exports[`runs a net with "xor" join 2`] = `
{
  "conditions": [
    {
      "marking": 1,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "choice",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
  ],
  "tasks": [
    {
      "generation": 0,
      "name": "initial_task",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_b",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_c",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "xor-join",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with "xor" join 3`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 1,
      "name": "choice",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "initial_task",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_a",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_b",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_c",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "xor-join",
      "parent": null,
      "state": "started",
    },
  ],
}
`;

exports[`runs a net with "xor" join 4`] = `
{
  "conditions": [
    {
      "marking": 0,
      "name": "start",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "choice",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "end",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_a->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 1,
      "name": "implicit:task_b->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "marking": 0,
      "name": "implicit:task_c->finish_task",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
  ],
  "tasks": [
    {
      "generation": 1,
      "name": "initial_task",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_a",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 1,
      "name": "task_b",
      "state": "completed",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "task_c",
      "state": "disabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
    {
      "generation": 0,
      "name": "finish_task",
      "state": "enabled",
      "workflowId": "workflow-1",
      "workflowName": "xor-join",
    },
  ],
  "workItems": [],
  "workflows": [
    {
      "context": undefined,
      "id": "workflow-1",
      "name": "xor-join",
      "parent": null,
      "state": "started",
    },
  ],
}
`;
